src/rgb_led.rs

```diff
@@ -0,0 +1,230 @@
+// Modified from https://github.com/fkohlgrueber/esp32c3-idf-led-example
+use std::{ffi::c_void, thread, time::Duration};
+
+use esp_idf_svc::sys::{
+    esp_err_t, gpio_num_t, rmt_carrier_level_t_RMT_CARRIER_LEVEL_HIGH,
+    rmt_carrier_level_t_RMT_CARRIER_LEVEL_LOW, rmt_channel_t, rmt_config_t,
+    rmt_config_t__bindgen_ty_1, rmt_mode_t_RMT_MODE_TX, rmt_tx_config_t, ESP_OK,
+};
+
+const WS2812_T0H_NS: u32 = 350;
+const WS2812_T0L_NS: u32 = 1000;
+const WS2812_T1H_NS: u32 = 1000;
+const WS2812_T1L_NS: u32 = 350;
+//const WS2812_RESET_US: u32 = 280;
+
+#[derive(Debug)]
+pub struct EspError {
+    inner: esp_err_t,
+}
+
+impl std::fmt::Display for EspError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "esp error code: {}", self.inner)
+    }
+}
+
+impl std::error::Error for EspError {}
+
+fn esp_res(err_code: esp_err_t) -> Result<(), EspError> {
+    if err_code == ESP_OK {
+        Ok(())
+    } else {
+        Err(EspError { inner: err_code })
+    }
+}
+
+pub type Led = LedStrip<1>;
+
+#[derive(Debug)]
+pub struct LedStrip<const NUM_LEDS: usize> {
```

Comment:
Why is it called LedStrip? It's like an array of LEDs?

Reply:
Yes, this is a driver for a type of addressable LED strips. The number of LEDs in the strip is given as a generic const. The ESP32 board only has 1 LED, but this driver is written in a way that would make it easy to expand to more than 1 if needed.